mobile/fula.go

```diff
@@ -48,66 +64,184 @@ type DirEntry struct {
+	return f.ds.Has(ctx, toDatastoreKey(l))
+}
+
+func toDatastoreKey(l ipld.Link) datastore.Key {
+	return datastore.NewKey(l.Binary())
+}
+
+func (f *Fula) Fetch(ctx context.Context, from peer.ID, l ipld.Link) error {
+	if exists, err := f.Has(ctx, l); err != nil {
 		return err
+	} else if exists {
+		return nil
 	}
+	resps, errs := f.gx.Request(ctx, from, l, selectorparse.CommonSelector_ExploreAllRecursively)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case resp, ok := <-resps:
+			if !ok {
+				return nil
+			}
+			log.Infow("synced node", "node", resp.Node)
+		case err, ok := <-errs:
+			if !ok {
+				return nil
+			}
+			log.Warnw("sync failed", "err", err)
+		}
+	}
+}
+
+func (f *Fula) TestGraphExchange(pid string, mas string, link string) (string, error) {
+	ctx := context.Background()
+
+	ma, err := multiaddr.NewMultiaddr(mas)
+	if err != nil {
+		return "", err
+	}
+	addrs := []multiaddr.Multiaddr{ma}
+	id, err := peer.Decode(pid)
+	if err != nil {
+		return "", err
+	}
+	f.h.Peerstore().AddAddrs(id, addrs, peerstore.PermanentAddrTTL)
+	if err = f.h.Connect(ctx, peer.AddrInfo{ID: id, Addrs: addrs}); err != nil {
+		return "", err
+	}
+	fmt.Printf("Connected to host %s", id)
+
+	rcid, err := cid.Decode(link)
+	if err != nil {
+		return "", err
+	}
+	rootLink := cidlink.Link{Cid: rcid}
 
-	hasAdder := len(peerAddr.Addrs) != 0
-	if hasAdder {
-		if manet.IsIPLoopback(peerAddr.Addrs[0]) {
-			return errors.New("cant Use loop back")
+	if err := f.Fetch(ctx, id, rootLink); err != nil {
+		return "", err
+	}
+
+	if exists, err := f.Has(ctx, rootLink); err != nil {
+		return "", err
+	} else if !exists {
+		return "", errors.New("expected host to have fetched the entire sample DAG")
+	} else {
+		fmt.Printf("%s successfully fetched:\n    link: %s\n    from %s\n", f.h.ID(), rootLink, id)
+		n, err := f.Load(ctx, rootLink, basicnode.Prototype.Any)
+		if err != nil {
+			return "", err
 		}
-		if manet.IsIPUnspecified(peerAddr.Addrs[0]) {
-			return errors.New("not Specified")
+		var buf bytes.Buffer
+		if err := dagjson.Encode(n, &buf); err != nil {
+			return "", err
 		}
+
+		res := fmt.Sprintf("    content: %s\n", buf.String())
+		fmt.Printf(res)
+		return res, nil
 	}
+}
 
-	if err = f.Node.Connect(f.ctx, *peerAddr); err != nil {
-		log.Error("peer router failed ", err.Error())
+func (f *Fula) AddBox(id peer.ID, addrs []multiaddr.Multiaddr) error {
+
+	if err := f.h.Connect(f.ctx, peer.AddrInfo{ID: id, Addrs: addrs}); err != nil {
+		log.Error("peer connect failed ", err.Error())
 		return err
 	}
 
-	f.mybox = append(f.mybox, *peerAddr)
+	f.mybox = append(f.mybox, peer.AddrInfo{ID: id, Addrs: addrs})
 	return nil
 }
 
 func (f *Fula) NewStream() (network.Stream, error) {
+	// @TODO
+	return nil, nil
+}
 
-	peerID, err := f.GetBox()
-	if err != nil {
-		return nil, err
-	}
-	log.Debug("box found", peerID)
+func (f *Fula) HostID() peer.ID {
+	return f.h.ID()
```

Comment:
General golang style comment: Avoid over-refactoring.

This is different from languages like java or JS. The rule of thumb I generally follow is that if a function is only being used in one place or it is quite short to write out, then i'd avoid refactoding it into receiver functions.

Reply:
The reason for the `HostID()` and `HostAddr()` functions is to prevent exposing `fula.h` (which is the libp2p host inside fula client). Since we need to have access to peer ID and MultiAddr in tests, I added these two receiver functions.
Is keeping the libp2p host private the right way to go @masih ? or should we make the `fula.h` public?

Reply:
For testing, you can expose `h` in the same package by defining a go file that has _test prefix in the same package and move the exported functions there. that means we would only expose those in testing and document that they are only used by testing in an executable way :)

Reply:
> keeping the libp2p host private

A general practive is to pass an already instantiated host to packages, but each package then owns the reference passed to it instead of making it public

Reply:
> For testing, you can expose `h` in the same package by defining a go file that has _test prefix in the same package and move the exported functions there. that means we would only expose those in testing and document that they are only used by testing in an executable way :)

Does it include in the package exports then? because we need it to be exported so that `gomobile` will include it in the resulting Java/Obj-c package